docs/EVE-IMAGE-REPRODUCIBILITY.md

```diff
@@ -0,0 +1,286 @@
+any non-reproducible source slipping in.
+
+Further, the EVE project is exploring additional gates. These include:
+
+* Contributors certifying on each Pull Request that they have not used non-reproducible sources.
+* Regular automated audits of all `pkg/` sources.
+* CI scans of `pkg/` sources.
+
+## Linux Packages
+
+The primary source for many software packages in Linux is the OS packaging system. All of the OCI images used in EVE are based on
+[Alpine](https://hub.docker.com/_/alpine), which uses the [apk package manager](https://wiki.alpinelinux.org/wiki/Alpine_Package_Keeper).
+
+For example, to install bash:
+
+```sh
+apk add bash
+```
+
+Alpine packages creates unique challenges for reproducibility. Specifically:
+
+1. The version of the package changes when changing the alpine version. `apk add bash` can give entirely different bits when running on `alpine:3.15` vs `alpine:3.16`.
+1. The version of the package can change even _without_ changing the alpine version. `apk add bash` can give a different version running today on `alpine:3.16` vs yesterday or tomorrow on the same base `alpine:3.16`.
+
+It is possible to mitigate, but not completely solve, the above issue by adding a specific version:
+
+```sh
+apk add bash=5.1.16-r2
+```
+
+The above does not completely solve it, as sometimes alpine packages change their bits _even for the same version_. Thus, running
+`apk add bash=5.1.16-r2` on the same base `alpine:3.16` can give different bits on different days. Alpine packages are changed "under the covers"
+when bugs are found or updates are provided.
+
+In order to resolve this issue, both the external/upstream `linuxkit/*` packages and the `lfedge/eve-*` packages in this repository follow the same principles.
+
+1. Create an alpine "cache" image: `linuxkit/alpine` and `lfedge/eve-alpine`. These are used as the base `FROM` for downstream images.
+1. All packages included in the composed EVE image **must** use `FROM lfedge/eve-alpine` or, in the case of linuxkit packages, `FROM linuxkit/alpine`, in their `Dockerfile`. `FROM alpine` is strictly forbidden.
+1. The cache images source from upstream alpine, download all necessary apk packages needed by downstream, `lfedge/eve-alpine` for EVE packages and `linuxkit/alpine` for linuxkit packages, and cache them.
+1. When downstream images need to install packages, they run `apk add package` which sources it solely from the cache in the cache image.
+
+The cache images - `lfedge/eve-alpine` and `linuxkit/alpine` - act as pinning cache, with precise versions and bits of each package.
+
+Only when the cache is regenerated, i.e. when `lfedge/eve-alpine` or `linuxkit/alpine` is rebuilt, can these change.
+
+Further, the caches maintain hashes of the local cache contents, to ensure that any changes are known and captured.
+
+To make this more concrete, we will walk through a specific example, `lfedge/eve-dom0-ztools`, which is based in [../pkg/dom0-ztools](../pkg/dom-ztools/).
```

Comment:
Bad link; dom-ztools should be dom0-ztools

Reply:
Oops. Fixing.